### ISSUE TITLE
Slow Zookeeper Pod Recovery (in PKS)

### ISSUE THREAD

**Issue by RaulGracia** [pravega/zookeeper-operator#72]
**Problem description**
In PR #69, the membership management of a Zookeeper cluster has been fixed. That is, at this point a Zookeeper ensemble has a consistent configuration along time, even in the presence of pod restarts. 

However, during the investigation of #69, another problem popped up: _the long time it takes for Zookeeper to recover from pod deletions/restarts in PKS_. Zookeeper is a critical dependency for Pravega and Bookkeeper, so we have examples of the impact of this problem in numerous reported issues: https://github.com/pravega/pravega/issues/3942, https://github.com/pravega/pravega/issues/3836, https://github.com/pravega/pravega/issues/3783, https://github.com/pravega/pravega/issues/3954. That is, among other symptoms, having an unhealthy Zookeeper service running in the cluster for too long may impact the Controller (e.g., unexpected restarts), Bookkeeper (e.g., inability to access metadata), Segment Store (e.g., problems with container recoveries, container assignments), as well as client applications (e.g., readers may exhaust their retries).

In PKS, this problem can be consistently reproduced. This is an example:

1) We have a complete Pravega deployment in PKS:
```
λ kubectl.exe get pods
NAME                                                    READY   STATUS    RESTARTS   AGE
alert-seagull-nfs-client-provisioner-5d7745fd84-4dhpn   1/1     Running   0          51m
benchmark-pod-3                                         1/1     Running   0          41m
original-ibex-nfs-client-provisioner-86c46b5d6d-8d5kl   1/1     Running   0          56m
pravega-bookie-0                                        1/1     Running   0          2m45s
pravega-bookie-1                                        1/1     Running   0          2m45s
pravega-bookie-2                                        1/1     Running   0          2m45s
pravega-operator-554d769d4c-gbdj7                       1/1     Running   0          43m
pravega-pravega-controller-d599b4fd6-wj5cd              1/1     Running   0          2m45s
pravega-pravega-segmentstore-0                          1/1     Running   0          2m45s
zookeeper-0                                             1/1     Running   0          5m57s
zookeeper-1                                             1/1     Running   0          5m17s
zookeeper-2                                             1/1     Running   0          4m36s
zookeeper-3                                             1/1     Running   0          4m9s
zookeeper-4                                             1/1     Running   0          3m40s
zookeeper-operator-566694c7ff-stzkn                     1/1     Running   0          59m
```
2) I have started a workload to verify that everything works fine:
```
/bin/pravega-benchmark -controller tcp://172.25.194.11:9090 -producers 1 -segments 1 -size 100 -events 100 -time 300 -stream myStream
...
[epollEventLoopGroup-4-1] INFO io.pravega.client.segment.impl.SegmentOutputStreamImpl - Connection setup complete for writer efc88ec9-d8a7-436b-a966-02169cd8140c
     501 records Writing,     100.1 records/sec,   0.01 MB/sec,     8.2 ms avg latency,    87.0 ms max latency
     501 records Writing,     100.0 records/sec,   0.01 MB/sec,     5.0 ms avg latency,    13.0 ms max latency
     501 records Writing,     100.0 records/sec,   0.01 MB/sec,     5.2 ms avg latency,    23.0 ms max latency
     501 records Writing,     100.0 records/sec,   0.01 MB/sec,     4.7 ms avg latency,    11.0 ms max latency
     501 records Writing,     100.1 records/sec,   0.01 MB/sec,     6.4 ms avg latency,    58.0 ms max latency
```
3) Then, we delete 2 out of 5 Zookeeper pods, expecting the service to automatically recover them:
```
λ kubectl delete pods zookeeper-2 zookeeper-0 --grace-period=0 --force
warning: Immediate deletion does not wait for confirmation that the running resource has been terminated. The resource may continue to run on the cluster indefinitely.
pod "zookeeper-2" force deleted
pod "zookeeper-0" force deleted
```
4) Zookeeper pods and, consequently, Controller pods keep restarting while the service is trying to recover Zookeeper pods:
```
λ kubectl.exe get pods
NAME                                                    READY   STATUS    RESTARTS   AGE
alert-seagull-nfs-client-provisioner-5d7745fd84-4dhpn   1/1     Running   0          57m
benchmark-pod-3                                         1/1     Running   0          46m
original-ibex-nfs-client-provisioner-86c46b5d6d-8d5kl   1/1     Running   0          61m
pravega-bookie-0                                        1/1     Running   0          8m11s
pravega-bookie-1                                        1/1     Running   0          8m11s
pravega-bookie-2                                        1/1     Running   0          8m11s
pravega-operator-554d769d4c-gbdj7                       1/1     Running   0          49m
pravega-pravega-controller-d599b4fd6-wj5cd              0/1     Running   2          8m11s
pravega-pravega-segmentstore-0                          1/1     Running   0          8m11s
zookeeper-0                                             0/1     Running   4          4m50s
zookeeper-1                                             1/1     Running   0          10m
zookeeper-2                                             0/1     Running   4          4m51s
zookeeper-3                                             1/1     Running   0          9m35s
zookeeper-4                                             1/1     Running   0          9m6s
zookeeper-operator-566694c7ff-stzkn                     1/1     Running   0          64m
```

5) The restarted Zookeeer pods get stuck in the very same point as in the rest of tests we have done so far (see `Connecting to Zookeeper zookeeper-client:2181`):
```
λ kubectl.exe logs zookeeper-0
...
Name:      zookeeper-headless.default.svc.cluster.local
Address 1: 172.25.194.10 172-25-194-10.zookeeper-client.default.svc.cluster.local
Address 2: 172.25.194.9 172-25-194-9.zookeeper-client.default.svc.cluster.local
Address 3: 172.25.194.5 172-25-194-5.zookeeper-client.default.svc.cluster.local
+ [[ 0 -eq 1 ]]
+ set -e
+ set +e
++ zkConnectionString
++ set +e
++ nslookup zookeeper-client
++ [[ 0 -eq 1 ]]
++ set -e
++ echo zookeeper-client:2181
+ ZKURL=zookeeper-client:2181
+ set -e
++ java -Dlog4j.configuration=file:/conf/log4j-quiet.properties -jar /root/zu.jar get-all zookeeper-client:2181
Connecting to Zookeeper zookeeper-client:2181
```

After a period of time between 10 minutes to 15 minutes, the system comes back to stability.

---

However, _the same experiment in GKE has a different outcome_: I measured the time from the deletion of pods to their total recovery in `1min 10secs`. In this case, the recovery was fast enough to do not induce any Controller restart:

```
raul_gracia@cloudshell:~ (pravega-dev)$ kubectl get pods
NAME                                        READY   STATUS    RESTARTS   AGE
coiling-shark-nfs-server-provisioner-0      1/1     Running   0          5h54m
pravega-bookie-0                            1/1     Running   1          6m22s
pravega-bookie-1                            1/1     Running   0          6m22s
pravega-bookie-2                            1/1     Running   0          6m22s
pravega-operator-687bbf897b-82z69           1/1     Running   0          155m
pravega-pravega-controller-f4dfc887-9gzb4   1/1     Running   0          6m22s
pravega-pravega-segmentstore-0              1/1     Running   0          6m22s
zookeeper-0                                 1/1     Running   0          72s
zookeeper-1                                 1/1     Running   0          8m57s
zookeeper-2                                 1/1     Running   0          23s
zookeeper-3                                 1/1     Running   0          7m49s
zookeeper-4                                 1/1     Running   0          7m5s
zookeeper-operator-5fb6d7cf4d-tlpsr         1/1     Running   0          5h```
```

---

Observations of this issue so far:
- The issue is reproducible in PKS, not in GKE.
- While in PKS it may take minutes for Zookeeper to stabilize, in general we observed that this is a transient problem and the system recovers.
- In PKS, the slow Zookeeper pod restarts occurs _the first time we delete Zookeeper pods_. Subsequent Zookeeper pod deletions are recovered fast, without problems. This leads us to think that after the Zookeeper cluster recovers from the first deletion, the resolution for the name `zookeeper-client:2181` is cached (this is the point in which Zookeeper pods get stuck when trying to recover). If the name resolution in PKS was slow due to some DNS problem, this may explain why subsequent pod deletions are handled faster. 

**Problem location**
PKS networking configuration? DNS?

**Suggestions for an improvement**
- Try with a longer readiness probe interval in Zookeeper pods.
- Understand and potentially fix the DNS architecture in our PKS clusters.

**Comment by EronWright** [comment]
I observe that the startup script is connecting to ZK via the `zookeeper-client` service, which routes thru a K8 load balancer (and may have platform-specific behavior).  Maybe the script should use the headless service, to let the ZK CLI figure out which member to connect to.

**Comment by pbelgundi** [comment]
Tried using the zk-headless service instead of zk-client service for connecting to zk-server and still noticed the same hang on PKS machines as in case of the zk-client.

**Comment by RaulGracia** [comment]
We have also tested using a larger readiness probe interval without success.

**Comment by Tristan1900** [comment]
One temporary fix has been tested to work by @RaulGracia and me. The fix is to modify this [value](https://github.com/pravega/zookeeper-operator/blob/master/docker/zu/src/main/java/io/pravega/zookeeper/Zookeeper.kt#L20) to be 1 and the restarted pod will be able to connect quickly. On `Nightshift` and `Frost`, it can reconnect within 4 minutes, on `Shipyard` it takes longer but still better than before. So for now, I think this is probably going to mitigate the problem.

The reason behind this is unclear and still needs investigation. But having a shorter timeout will throw an exception and let the pod restart and try to connect to the service again.  Somehow this can reduce the time waiting to connect.

**Comment by RaulGracia** [comment]
As an update, this issue only occurs in PKS, either on-premise or deployed in AWS. Conversely, the issue does not appear in EKS or GKE and pods recover within 1-2 minutes.

**Comment by RaulGracia** [comment]
This has been fixed in this PR #93.
